doc/sphinx/proofs/writing-proofs/rewriting.rst

```diff
@@ -213,89 +217,135 @@ simply :g:`t=u` dropping the implicit type of :g:`t` and :g:`u`.
+   .. exn:: Section variable @ident occurs implicitly in global declaration @qualid present in hypothesis @ident.
+            Section variable @ident occurs implicitly in global declaration @qualid present in the conclusion.
 
       Raised when the variable is a section variable with indirect
       dependencies in the goal.
+      If :n:`@ident` is a section variable, it must not have any
+      indirect occurrences in the goal, i.e. no global declarations
+      implicitly depending on the section variable may be present in the
+      goal.
 
+.. tacn:: simple subst
+   :undocumented:
 
-.. tacn:: stepl @term
-   :name: stepl
+.. tacn:: stepl @one_term {? by @ltac_expr }
 
-   This tactic is for chaining rewriting steps. It assumes a goal of the
-   form :n:`R @term @term` where ``R`` is a binary relation and relies on a
+   For chaining rewriting steps. It assumes a goal in the
+   form :n:`R @term__1 @term__2` where ``R`` is a binary relation and relies on a
    database of lemmas of the form :g:`forall x y z, R x y -> eq x z -> R z y`
-   where `eq` is typically a setoid equality. The application of :n:`stepl @term`
-   then replaces the goal by :n:`R @term @term` and adds a new goal stating
-   :n:`eq @term @term`.
+   where `eq` is typically a setoid equality. The application of :n:`stepl @one_term`
+   then replaces the goal by :n:`R @one_term @term__2` and adds a new goal stating
+   :n:`eq @one_term @term__1`.
 
-   .. cmd:: Declare Left Step @term
+   If :n:`@ltac_expr` is specified, it is applied to the side condition.
 
-      Adds :n:`@term` to the database used by :tacn:`stepl`.
+   .. cmd:: Declare Left Step @one_term
+
+      Adds :n:`@one_term` to the database used by :tacn:`stepl`.
 
    This tactic is especially useful for parametric setoids which are not accepted
    as regular setoids for :tacn:`rewrite` and :tacn:`setoid_replace` (see
    :ref:`Generalizedrewriting`).
 
-   .. tacv:: stepl @term by @tactic
-
-      This applies :n:`stepl @term` then applies :token:`tactic` to the second goal.
+   .. tacn:: stepr @one_term {? by @ltac_expr }
 
-   .. tacv:: stepr @term by @tactic
-      :name: stepr
-
-      This behaves as :tacn:`stepl` but on the right-hand-side of the binary
-      relation. Lemmas are expected to be of the form
+      This behaves like :tacn:`stepl` but on the right hand side of the binary
+      relation. Lemmas are expected to be in the form
       :g:`forall x y z, R x y -> eq y z -> R x z`.
 
-   .. cmd:: Declare Right Step @term
+   .. cmd:: Declare Right Step @one_term
 
        Adds :n:`@term` to the database used by :tacn:`stepr`.
 
+Rewriting with definitional equality
+------------------------------------
+
+.. tacn:: change @conversion {? @occurrences }
+
+   .. insertprodn conversion conversion
+
+   .. prodn::
+      conversion ::= @one_term {? {? at @occs_nums } with @one_term }
```

Comment:
OK then, let's not splice `conversion`. But then, I propose to transform it so that it appears as two separate cases, like this:
```suggestion
      conversion ::= @one_term__to
      | @one_term__from {? at @occs_nums } with @one_term__to
```

Another option, with a single case is this:
```suggestion
      conversion ::= {? @one_term__from {? at @occs_nums } with } @one_term__to
```

The goal of this change being to highlight that (contrary to `replace`), the `one_term__to` must always be provided and it is the `one_term__from` that may be omitted (in which case the whole goal is replaced).

Reply:
The only way to do this is to remove the `insertprodn`, which means it won't be updated by `doc_grammar`.  I'll splice it.

Reply:
What!? No!

I was just suggesting an edit to add to `common.edit_mlg`, like we used to do all the time.

Reply:
Can you be more specific?  No to splicing it?  That was your original suggestion.

What edit do you want?  Specifying subscripts from `common.edit_mlg` is not a good idea, `doc_grammar` is really not set up to support that.

Reply:
OK my mistake. I misunderstood your comment and didn't realize that you cannot specify subscripts in a `prodn` (would it be hard to add?). It looks good like this.

Reply:
At this point I'm reluctant to add anything to `doc_grammar` unless it's really necessary or quite simple.  It might take a couple days to do this.

For various reasons, the syntax updates are taking a very long time (long delays), which is discouraging.  It indicates that any significant improvements to Coq are likely to take a very long time to complete and really undercuts the motivation to undertake new improvements.

Reply:
There ought to be an aphorism something like "Gratification deferred is gratification denied".

Reply:
Or let me put it this way: I don't want to make such a change a prerequisite to completing this PR.

Reply:
It was just a suggestion, not a request, and certainly not a prerequisite for this PR.

However, since you complain about your doc PRs taking a very long time to merge, I'll respond. I've spent considerable more time reviewing PRs (yours) this year that I had ever before. I do this because I see the value but I don't have unlimited time to allocate to this task, so it takes time and there are delays. You're opening PRs containing very large changes. This is not the good strategy if you want rapid reviews and merging. Your syntax updates PRs are usually not in a great shape when you open them (descriptions do not match anymore with the documented syntax) so it's normal that many feedback cycles are needed to get them ready. (And of course, the time zone difference between the two of us is also significantly delaying these feedback cycles even when I'm the most available, like during the first lockdown.)

Reply:
My comment was not meant as a complaint directed at you.  I'm sorry if my wording gave that impression.  Like everyone else, you're busy with many other things and your participation on any given effort is voluntary.  Your detailed reviews have been indispensable and very much appreciated.

I agree it's a challenging process getting these PRs right.  I'm open to suggestions to make it go more smoothly--though I think much of the challenge is probably inherent rather than due to our use of an inadequate process.

The current form of `doc_grammar` is set up to update an entire rst file, which makes it a natural increment for PRs.  That's too much to do at once for the tactics chapter (3000 lines), SSR (5800 lines) and maybe too much for the rewriting chapter (800 lines).  We could take an approach in which we do full updates on, say, the first 400 lines.  When that part is ready to commit, I can create a second branch that keeps all the changes.  Then, on the first branch, I remove all the changes made in `rewriting.rst` outside those 400 lines before we commit the changes (with possible exceptions in some cases, such as for material moved from the end into the 400 lines).   
 Then do the rest of the file in a new PR.  Do you want to try that for this PR?

Reply:
>much of the challenge is probably inherent

Agreed.

>Do you want to try that for this PR?

:+1: Since I never managed to push my review beyond the documentation of `change`, it would make sense to postpone the rest to a later PR. This would also help ensure that this PR gets merged in time for 8.13.0.

Reply:
OK, we'll make this one cover everything before `Performing computations`.